presto-iceberg/src/main/java/com/facebook/presto/iceberg/procedure/RemoveOrphanFiles.java

```diff
@@ -0,0 +1,226 @@
+        Table icebergTable = IcebergUtil.getIcebergTable(metadata, clientSession, schemaTableName);
+
+        Set<String> processedManifestFilePaths = new HashSet<>();
+        ImmutableSet.Builder<String> validMetadataFileNames = ImmutableSet.builder();
+        ImmutableSet.Builder<String> validDataFileNames = ImmutableSet.builder();
+
+        for (Snapshot snapshot : icebergTable.snapshots()) {
+            if (snapshot.manifestListLocation() != null) {
+                validMetadataFileNames.add(extractFileName(snapshot.manifestListLocation()));
+            }
+
+            for (ManifestFile manifest : snapshot.allManifests(icebergTable.io())) {
+                if (!processedManifestFilePaths.add(manifest.path())) {
+                    // Already read this manifest
+                    continue;
+                }
+
+                validMetadataFileNames.add(extractFileName(manifest.path()));
+                try (ManifestReader<? extends ContentFile<?>> manifestReader = readerForManifest(icebergTable, manifest)) {
+                    for (ContentFile<?> contentFile : manifestReader) {
+                        validDataFileNames.add(extractFileName(contentFile.path().toString()));
+                    }
+                }
+                catch (IOException e) {
+                    throw new PrestoException(ICEBERG_FILESYSTEM_ERROR, "Unable to list manifest file content from " + manifest.path(), e);
+                }
+            }
+        }
+
+        metadataFileLocations(icebergTable, false).stream()
+                .map(this::extractFileName)
+                .forEach(validMetadataFileNames::add);
+
+        statisticsFilesLocations(icebergTable).stream()
+                .map(this::extractFileName)
+                .forEach(validMetadataFileNames::add);
+
+        validMetadataFileNames.add("version-hint.text");
+
+        // Remove unused metadata and data files older than 3 days by default
+        long expiration = olderThan == null ?
+                System.currentTimeMillis() - DAYS.toMillis(REMOVE_UNUSED_FILES_OLDER_THAN_IN_DAYS) :
+                olderThan.isLegacyTimestamp() ? olderThan.getMillisUtc() : olderThan.getMillis();
+        scanAndDeleteInvalidFiles(icebergTable, clientSession, schemaTableName, expiration, validDataFileNames.build(), "data");
+        scanAndDeleteInvalidFiles(icebergTable, clientSession, schemaTableName, expiration, validMetadataFileNames.build(), "metadata");
+    }
+
+    private void scanAndDeleteInvalidFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expiration, Set<String> validFiles, String subFolder)
+    {
+        try {
+            List<String> filesToDelete = new ArrayList<>();
+            FileSystem fileSystem = getFileSystem(session, this.hdfsEnvironment, schemaTableName, new Path(table.location()));
+            Path subFolderPath = new Path(table.location(), subFolder);
```

Comment:
You should instead use the table properties `write.data.path` and `write.metadata.path` as they are not guaranteed to be `data` and `metadata`: https://iceberg.apache.org/docs/1.5.2/configuration/#write-properties

Reply:
Referring to Iceberg's implementation, I extracted two methods `IcebergUtil.metadataLocation(icebergTable)` and `IcebergUtil.dataLocation(icebergTable)`. They took into account the relevant table properties to obtain metadata location and data location of the target Iceberg table. Please take a look when available, thanks!

Besides, we currently haven't supported these additional table properties in presto, so temporarily unable to add test cases to show the scenarios that table properties affecting the default value of metadata location and data location. We can add these test cases in 'TestRemoveOrphanFilesPocedure' once introducing the table properties.

Reply:
> Besides, we currently haven't supported these additional table properties in presto, so temporarily unable to add test cases to show the scenarios that table properties affecting the default value of metadata location and data location.

One way to get around this might be to create a new `Table` instance through the Java API with the properties set, and then call the `register_table` procedure?

Reply:
Thanks, this is feasible. I have added test cases for both `HADOOP` and `HIVE` catalog separately as there are some differences between them. `HADOOP` catalog do not support table property `WRITE_METADATA_LOCATION` on table creation, while the implementation of `HIVE` catalog in presto is based on self implemented `HiveTableOperations` rather than using Iceberg lib's `HiveCatalog`, so the behavior of `createTable()/dropTable()/loadTable()` are different between the two. Please take a look when available.